lib/bitrange.h

```diff
@@ -51,4 +58,714 @@ class bitranges {
+     *             Space:               [o o o o o o o o o o]
+     *             Big endian range:    --> [2 3 4]
+     *             Little endian range:     [7 6 5] <--------
+     * We couldn't switch between those numberings without knowing that the
+     * space has 10 elements.
+     *
+     * Note that this operation may result in integer overflow when dealing with
+     * very large ranges. It's generally safe for ZeroToMax as long as the space
+     * size is not also near INT_MAX. It's never safe for MinToMax.
+     *
+     * @tparam DestOrder  The endian order to convert to.
+     * @param spaceSize  The size of the space this range is embedded in.
+     * @return this range, but converted to the specified endian ordering.
+     */
+    template <Endian DestOrder>
+    HalfOpenRange<Unit, DestOrder> toOrder(int spaceSize) const {
+        if (DestOrder == Order) return HalfOpenRange<Unit, DestOrder>(lo, hi);
+        switch (DestOrder) {
+            case Endian::Network:
+            case Endian::Little:
+                return HalfOpenRange<Unit, DestOrder>(spaceSize - hi, spaceSize - lo);
+        }
+        BUG("Unexpected ordering");
+    }
+
+    /**
+     * Convert this range to the smallest enclosing range with the specified unit.
+     *
+     * Conversion from bytes to bits is exact, but conversion from bits to bytes
+     * is lossy and may cause the size of the range to grow if the bits aren't
+     * byte-aligned. If that would be problematic, use `isLoAligned()` and
+     * `isHiAligned()` to check before converting, or just check that converting
+     * the result back to bits yields the original range.
+     *
+     * This operation will result in integer overflow when dealing with
+     * ZeroToMax or MinToMax-sized ranges.
+     *
+     * @tparam DestUnit  The unit to convert to.
+     * @return this range, but converted to the specified unit.
+     */
+    template <RangeUnit DestUnit>
+    HalfOpenRange<DestUnit, Order> toUnit() const {
+        if (DestUnit == Unit) return HalfOpenRange<DestUnit, Order>(lo, hi);
+        if (empty()) return HalfOpenRange<DestUnit, Order>();
+        switch (DestUnit) {
+            case RangeUnit::Bit:
+                return HalfOpenRange<DestUnit, Order>(lo * 8, hi * 8);
+            case RangeUnit::Byte:
+                return HalfOpenRange<DestUnit, Order>(loByte(), nextByte());
+        }
+        BUG("Unexpected unit");
+    }
+
+    /// JSON serialization/deserialization.
+    void toJSON(JSONGenerator &json) const { BitRange::rangeToJSON(json, lo, hi); }
+    static HalfOpenRange fromJSON(JSONLoader &json) {
+        return HalfOpenRange(BitRange::rangeFromJSON(json));
+    }
+
+    /// Total ordering, first by lo, then by hi.
+    bool operator<(const HalfOpenRange &other) const {
+        if (lo != other.lo) return lo < other.lo;
+        return hi < other.hi;
+    }
+
+    friend size_t hash_value(const HalfOpenRange &r) {
```

Comment:
Sorry, I missed this previously... While we are here, can we add `std::hash` / `Util::Hasher` specializations? Also, `Utils::hash_combine` combines `std::hash<T>`'s by default with `T` being `int` here, so I think it would be better to use `return Utils::Hash{}(r.lo, r.hi)` here, it would mix better (not sure it matters, but just in case).

Same for `ClosedRange` below.

Reply:
I think I've implemented what you've requested with the specializations. Please take a look and let me know if anything is missing.